lib/icinga/clusterevents-check.cpp

```diff
@@ -93,36 +93,111 @@ void ClusterEvents::ExecuteCheckFromQueue(const MessageOrigin::Ptr& origin, cons
+			return;
+		}
+
+		Checkable::ExecuteCommandProcessFinishedHandler = [checkableName, listener, sourceEndpoint, origin, params] (const Value& commandLine, const ProcessResult& pr) {
+			if (params->Get("command_type") == "check_command") {
+				Checkable::CurrentConcurrentChecks.fetch_sub(1);
+				Checkable::DecreasePendingChecks();
+			}
+
+			if (pr.ExitStatus > 3) {
+				Process::Arguments parguments = Process::PrepareCommand(commandLine);
+				Log(LogWarning, "ApiListener")
+					<< "Command for object '" << checkableName << "' (PID: " << pr.PID
+					<< ", arguments: " << Process::PrettyPrintArguments(parguments) << ") terminated with exit code "
+					<< pr.ExitStatus << ", output: " << pr.Output;
+			}
+
+			Dictionary::Ptr executedParams = new Dictionary();
+			executedParams->Set("execution", params->Get("source"));
+			executedParams->Set("host", params->Get("host"));
+			if (params->Contains("service"))
+				executedParams->Set("service", params->Get("service"));
+			executedParams->Set("exit", pr.ExitStatus);
+			executedParams->Set("output", pr.Output);
+			executedParams->Set("start", pr.ExecutionStart);
+			executedParams->Set("end", pr.ExecutionEnd);
+
+			if (origin->IsLocal()) {
+				ClusterEvents::ExecutedCommandAPIHandler(origin, executedParams);
+			} else {
+				Dictionary::Ptr executedMessage = new Dictionary();
+				executedMessage->Set("jsonrpc", "2.0");
+				executedMessage->Set("method", "event::ExecutedCommand");
+				executedMessage->Set("params", executedParams);
+
+				listener->SyncSendMessage(sourceEndpoint, executedMessage);
+			}
+		};
 	}
 
-	if (!listener->GetAcceptCommands()) {
+	if (!listener->GetAcceptCommands() && !origin->IsLocal()) {
 		Log(LogWarning, "ApiListener")
 				<< "Ignoring command. '" << listener->GetName() << "' does not accept commands.";
 
-		Host::Ptr host = new Host();
-		Dictionary::Ptr attrs = new Dictionary();
+		String output = "Endpoint '" + Endpoint::GetLocalEndpoint()->GetName() + "' does not accept commands.";
+
+		if (params->Contains("source")) {
+			Dictionary::Ptr executedParams = new Dictionary();
+			executedParams->Set("execution", params->Get("source"));
+			executedParams->Set("host", params->Get("host"));
+			if (params->Contains("service"))
+				executedParams->Set("service", params->Get("service"));
+			double now = Utility::GetTime();
+			executedParams->Set("exit", 126);
+			executedParams->Set("output", output);
+			executedParams->Set("start", now);
+			executedParams->Set("end", now);
 
-		attrs->Set("__name", params->Get("host"));
-		attrs->Set("type", "Host");
-		attrs->Set("enable_active_checks", false);
+			Dictionary::Ptr executedMessage = new Dictionary();
+			executedMessage->Set("jsonrpc", "2.0");
+			executedMessage->Set("method", "event::ExecutedCommand");
+			executedMessage->Set("params", executedParams);
 
-		Deserialize(host, attrs, false, FAConfig);
+			listener->SyncSendMessage(sourceEndpoint, executedMessage);
```

Comment:
I don't really like this boilerplate copied X times.

Evaluate how many lines of code (relative) per such message we'd save with a static function (not static method!) taking `params, `exit`, `output`, `start`, `end`, `listener`, `origin` and `sourceEndpoint`.